### DISCUSSION TITLE
Migration to Private Rainmaker Instance

### DISCUSSION THREAD

**Discussion by CrownedComedian** [espressif/arduino-esp32#9459] [Q&A]
Hello!

For convenience, I've been using the Arduino framework along with the ESP RainMaker library for Arduino to build and flash my project code.  I'm looking ahead to migrate my product to my own private instance of Rainmaker, which requires changes in the sdkconfig (private MQTT host url, and enable local control if possible).  In a normal ESP-IDF project, I could just use idf.py menuconfig to make these changes, but that's not an option in Arduino.  Is there an easy way to change my Arduino Rainmaker library configuration, or at least change the MQTT url and enable local control?

**Comment by Alli-Ekundayo** [comment]
I see! You're using the Arduino framework with the ESP RainMaker library and want to migrate to your own private instance of Rainmaker, which requires changes to the sdkconfig. Unfortunately, the Arduino environment doesn't provide a direct way to modify the sdkconfig like idf.py menuconfig does in a native ESP-IDF project.

However, there are a few workarounds you can try:

## - Modify the RainMaker library:
 You can try modifying the RainMaker library for Arduino to include your custom configuration. You'll need to edit the library's source code to change the default MQTT host URL and enable local control.
## - Use environment variables: 
You can pass environment variables to the RainMaker library during initialization, which might allow you to override the default configuration. Check the library's documentation to see if this is supported.
## - Create a custom wrapper: 
Write a custom wrapper around the RainMaker library that initializes it with your custom configuration. This would require creating a new Arduino library that wraps the RainMaker library and provides a custom initialization function.
## - Use a different library:
 If the above options are too complicated, you might consider using a different ESP32/ESP8266 library that provides more flexibility in configuration, such as the ESP-IDF Arduino wrapper or the PubSubClient library.

**Comment by dzungpv** [comment]
I am testing RM too, the solution I am using is using Arduino as IDF component https://espressif-docs.readthedocs-hosted.com/projects/arduino-esp32/en/latest/esp-idf_component.html , Then you can change whatever sdkconfig variable. And have many power feature of the IDF like flash encryption, secure boot... which does not exist in Arduino world.
